typing/rec_check.ml

```diff
@@ -1286,12 +1303,18 @@ let is_valid_recursive_expression idlist expr : sd option =
         (* The expression has known size *)
         let ty = expression expr Return in
         if Env.unguarded ty idlist = [] then Some Static else None
+     | Constant ->
+        (* The expression will not be pre-allocated, but the result cannot
+           contain recursive references *)
+        let ty = expression expr Return in
+        if Env.unguarded ty idlist = [] then Some Constant else None
```

Comment:
I think that we could share the code with the `Static` case above -- in my mind we have `Static of [block 
 | constant]` as two subcases of the same general idea.

```ocaml
  | (Static | Constant) as recursive_kind ->
        let ty = expression expr Return in
        if Env.unguarded ty idlist = [] then Some recursive_kind else None
```

we lose the comments, but the comments are not so helpful anyway at this point, they refer to the compilation scheme and not the code itself.